docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+# Entry-point based roles
+
+## We have a problem
+
+Each role consists of a set of tasks and variables read/set in some of the tasks.
+Some of the tasks can store their results into registered variables,
+e.g. store stdout of a docker pull command.
+Some of the tasks may be run only under certain condition that are generated
+outside of a role, e.g. pull images only if a deployment is containerized.
+Some of the tasks are run only if a certain variable is defined,
+e.g. add additional Kubelet arguments into the `node-config.yml`
+if `kubelet_additional_args` inventory variable is defined.
+
+Some roles work with a lot of variables that from a point of a role are uninvited citizens.
+Some of the variables are set inside a role, some of them are set outside of a role.
+The variables outside of a role are hard to track, hard to determine their value or
+a condition under which the variables are defined. These unknown facts make a role harder
+to maintain, harder to make a role closed with respect to variables and to make
+a role practically modular.
+
+## Proposal
+
+Make a role that conforms to the following conditions:
+* none of the variables read/set inside a role are global
+* all variables inside a role are `r_` (`r` standing for a role) and/or `l_`(`l` standing for a local) prefixed
+* all values are propagated to a role through `r_` prefixed variables only
+* all values that are set inside a role are read outside of a role through `r_` prefixed variables only
+* all available `r_` prefixed variables are documented and checked before a main role task is run
+* a table of `r_` prefixed variables in README.md is automatically generated
+
+### Terminilogy
+
+**Role variables**: All `r_` prefixed variables read/set inside a role
+
+**Local variables**: All `l_` prefixed variables set inside a role
+
+### Role properties
+
+**Solid API**: all variables that a role caller can set need to be well defined.
+It means each such variable must have the following information attached:
+* name
+* description
+* default value (if the variable is defined)
+* set of acceptable values
+* if it is required or if it is optional
+
+Well defined API allows automatically generated documentation and variable checks.
+
+**Free of outer variables**: only variables consisting the role API or defined locally are allowed.
```

Comment:
The `openshift_node` has a list of role variables that are set to the following non-role variables than do not have to be defined:
* openshift.node.dns_ip
* openshift.common.sdn_network_plugin_name
* openshift.common.ip
* openshift_node_min_tls_version
* openshift_node_cipher_suites
* openshift_disable_swap
* openshift_cloudprovider_aws_access_key
* openshift_cloudprovider_aws_secret_key
* openshift_cloudprovider_kind
* openshift_node_kubelet_args
* openshift.common.no_proxy
* openshift.node.debug_level
* openshift.common.https_proxy
* openshift.common.http_proxy
* openshift.common.system_images_registry
* openshift.node.env_vars
* openshift.docker.gte_1_10

All of them can be set to some default value. The question is which default value is the proper one. The only thing I am afraid of is duplication (and syncing) of default values among role variables and their corresponding non-role variables. On the other hand it is completely fine for a role variable to have a different value than its corresponding inventory variable. Since a default role variable should be always chosen from a point of view of a role and default inventory variable from a point of view of entire deployment.

So, if I can assume all of the variables are always going to be set to some default value, I no longer have to use the `set_fact: ...when: var is defined` sequence for all the above mentioned variables.